File: src/com/ptby/dynamicreturntypeplugin/typecalculation/ParameterTypeCalculator.java
package com.ptby.dynamicreturntypeplugin.typecalculation;

import com.intellij.psi.PsiElement;
import com.jetbrains.php.lang.psi.elements.PhpTypedElement;
import com.jetbrains.php.lang.psi.resolve.types.PhpType;
import com.ptby.dynamicreturntypeplugin.DynamicReturnTypeProvider;
import com.ptby.dynamicreturntypeplugin.index.ClassConstantAnalyzer;

public class ParameterTypeCalculator {
    private final ClassConstantAnalyzer classConstantAnalyzer;


    public ParameterTypeCalculator( ClassConstantAnalyzer classConstantAnalyzer1 ) {
        this.classConstantAnalyzer = classConstantAnalyzer1;
    }


    public ParameterType calculateTypeFromParameter( int parameterIndex, PsiElement[] parameters ) {
        if ( parameters.length <= parameterIndex ) {
            return new ParameterType( null );
        }

        PsiElement element = parameters[ parameterIndex ];
        if ( element instanceof PhpTypedElement ) {
            PhpType type = ( ( PhpTypedElement ) element ).getType();
            if ( !type.toString().equals( "void" ) ) {
                if ( type.toString().equals( "string" ) ) {
                    return new ParameterType( cleanClassText( element ) );
                } else if ( classConstantAnalyzer.verifySignatureIsClassConstant( type.toString() ) ) {
                    return new ParameterType( type.toString() );
                }

                String singleType = getTypeSignature( type );
                if ( singleType == null ) {
                    return new ParameterType( null );
                }

                if ( singleType.substring( 0, 1 ).equals( "\\" ) ) {
                    return new ParameterType( "#C" + singleType );
                }

                if ( singleType.length() < 3 ) {
                    return new ParameterType( null );
                }

                if ( typeContains( singleType, "#C\\" ) || typeContains( singleType, "#P#C\\" )  ) {
                    return new ParameterType( singleType.substring( 2 ) );
                }

                String calculatedType = singleType.substring( 3 );
                return new ParameterType( calculatedType );
            }
        }

        return new ParameterType( null );
    }


    private boolean typeContains( String singleType, String comparison ){
        String projectIdentifiedComparison = "#" + DynamicReturnTypeProvider.PLUGIN_IDENTIFIER_KEY + comparison;
        if( singleType.length() < projectIdentifiedComparison.length() ){
            return false;
        }

        return singleType.substring( 0, projectIdentifiedComparison.length() )
                         .equals( projectIdentifiedComparison );
    }


    private String getTypeSignature(  PhpType type ){
        String typeSignature = null;
        for( String singleType : type.getTypes() ){
            typeSignature = singleType;
        }

        return typeSignature;
    }


    private String cleanClassText( PsiElement element ) {
        String potentialClassName = element.getText().trim();
        if ( potentialClassName.equals( "" ) || potentialClassName.equals( "''") ) {
            return null;
        }

        return potentialClassName.replaceAll( "(\"|')", "" );
    }
}

File: src/com/ptby/dynamicreturntypeplugin/typecalculation/ParameterTypeCalculator.kt
package com.ptby.dynamicreturntypeplugin.typecalculation

import com.intellij.psi.PsiElement
import com.jetbrains.php.lang.psi.elements.PhpTypedElement
import com.jetbrains.php.lang.psi.resolve.types.PhpType
import com.ptby.dynamicreturntypeplugin.DynamicReturnTypeProvider
import com.jetbrains.php.lang.psi.elements.MethodReference
import com.jetbrains.php.lang.psi.elements.FunctionReference
import com.ptby.dynamicreturntypeplugin.signature_extension.matchesPhpClassConstantSignature

public class ParameterTypeCalculator() {


    public fun calculateTypeFromParameter(parameterIndex: Int, parameters: Array<PsiElement>): ParameterType {
        if (parameters.size() <= parameterIndex) {
            return ParameterType(null)
        }

        val element = parameters[parameterIndex]
        if (element is PhpTypedElement) {
            val elementType = (element).getType()
            if (elementType.toString() != "void") {
                if (elementType.toString() == "string") {
                    return ParameterType(cleanClassText(element))
                } else if ( elementType.toString().matchesPhpClassConstantSignature() ) {
                    return ParameterType(elementType.toString())
                }

                val singleType = getTypeSignature(elementType)
                        ?: return ParameterType(null)

                if (singleType.substring(0, 1) == "\\") {
                    return ParameterType("#C" + singleType)
                }

                if (singleType.length() < 3) {
                    return ParameterType(null)
                }

                if (typeContains(singleType, "#C\\") || typeContains(singleType, "#P#C\\")) {
                    return ParameterType(singleType.substring(2))
                }

                val calculatedType = singleType.substring(3)
                return ParameterType(calculatedType)
            }
        }

        return ParameterType(null)
    }


    private fun typeContains(singleType: String, comparison: String): Boolean {
        val projectIdentifiedComparison = "#" + DynamicReturnTypeProvider.PLUGIN_IDENTIFIER_KEY + comparison
        if (singleType.length() < projectIdentifiedComparison.length()) {
            return false
        }

        return singleType.substring(0, projectIdentifiedComparison.length()) == projectIdentifiedComparison
    }


    private fun getTypeSignature(phpType: PhpType): String? {
        var typeSignature: String? = null
        for (singleType in phpType.getTypes()) {
            typeSignature = singleType
        }

        return typeSignature
    }


    private fun cleanClassText(element: PsiElement): String? {
        val potentialClassName = element.getText().trim()
        if (potentialClassName == "" || potentialClassName == "''") {
            return null
        }

        return potentialClassName.replace("(\"|')".toRegex(), "").replace(":", "\\")
    }
}
